### PR TITLE
Re-write Rust SDK

### PR DESCRIPTION
**Pull request by sheepduke**
## This PR

Re-writes the OpenFeature Rust SDK based on the spec.

This PR brings the following major improvements compared to existing code:

* A singleton of `OpenFeature` is implemented as per [spec 1.1.1](https://openfeature.dev/specification/sections/flag-evaluation#requirement-111). The application author should call `get_client` to get an instance. No more manual creation.

* `EvaluationDetails` and `ResolutionsDetails` has been refined to use `enum` to store the reason and error code, per [spec](https://openfeature.dev/specification/types#resolution-details).

* Evaluation functions is now strongly typed. Meaning the user should call `get_int_value` to get an `i64` value. This is required by the [spec 1.3.1](https://openfeature.dev/specification/sections/flag-evaluation#requirement-131).

* Evaluation functions `get_xx_value` becomes `async` as suggested by the [spec 1.4.11](https://openfeature.dev/specification/sections/flag-evaluation#requirement-1411).

* Meaningful doc strings for public types and traits. Content selectively copied from the spec.

See Follow-up tasks for what has not been done.

### Related Issues

Fixes:
- #6
- #12 
- #13

### Notes

Why re-writing?

I tried to patch existing code but ended up re-writing it. The main reason was that too much code needed to be changed to make both spec and Rust compiler happy.

For example, the old `resolution` function takes a type parameter `T: Clone` that makes it non object safe, which breaks the requirements of singleton (`lazy_static`), `async` and more. The provider author can never put this `T` value in a cache because it has `Clone` constraint that implies `Sized` etc etc.

There are some other pain points and I cannot recall now. Still, the existing code inspired me in a lot of ways. Just I was a bit lazy to patch everything and re-writing would be easier.

### Follow-up Tasks

- Implement more evaluation functions `get_xx_details`.
- Implement hooks as per #5.
- Re-scan the spec and ensure everything is compliant.
- Use Clippy to guarantee the code quality (by running some selective rules).
- Address other issues in #9.

### CODE REVIEW

**Comment by beeme1mr** [comment]
Hey @sheepduke, could you please sign off your commits? It's a requirement from the CNCNF before we can merge.

To add your Signed-off-by line to every commit in this branch:

1. Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
2. In your local branch, run: `git rebase HEAD~12 --signoff`
3. Force push your changes to overwrite the branch: git push --force-with-lease origin main

**Review by beeme1mr — APPROVED** [review]
Looks good! As a follow-up, please update the readme to match the new package name and structure.

File: renovate.json
Comment (justinabrahms) [reply renovate.json:1]: Noting that this needs to be fixed before merge.

File: src/api/api.rs
Comment (justinabrahms) [inline src/api/api.rs:27]: Do we expect callers to call this, given there is a singleton that they should be using instead? If it's only for testing, we should wrap it in `#[cfg(test)]`

File: src/evaluation/context.rs
Comment (justinabrahms) [inline src/evaluation/context.rs:39]: Is there a reason the user would call this instead of the one above? I think if they don't care about the return value, then they just don't use it, which would simplify this API.

Or maybe the one above is intended to be non-mutating?

File: src/evaluation/details.rs
Comment (justinabrahms) [inline src/evaluation/details.rs:13]:
```suggestion
/// Reason for evaluation.
```

File: src/evaluation/details.rs
Comment (justinabrahms) [inline src/evaluation/details.rs:93]:
```suggestion
    General(String),
```

File: src/evaluation/field_value.rs
Comment (justinabrahms) [inline src/evaluation/field_value.rs:147]: Shouldn't `unwrap` handle the panic case (here and elsewhere)?

**Review by justinabrahms — CHANGES_REQUESTED** [review]
I'm biased b/c I wrote this code, but you might like the pattern we used in the java-sdk where each test is annotated with a specification that it covers. It helps uncover which areas we're not covering yet.

https://github.com/open-feature/java-sdk/blob/main/src/test/java/dev/openfeature/sdk/FlagEvaluationSpecTest.java#L69

File: src/api/api.rs
Comment (tcarrio) [reply src/api/api.rs:27]: If we can keep it private so only tests can manually create it that would probably be best. No reason to work outside the API. But if there is a use case please correct me here 🙂

File: renovate.json
Comment (sheepduke) [reply renovate.json:1]: Could you please kindly explain to me what does this file do? I googled it but did not see anywhere related. Plus no usage in the code...

File: src/evaluation/details.rs
Comment (sheepduke) [reply src/evaluation/details.rs:13]: I really appreciate your help of checking these small yet important details. :-)

File: src/evaluation/field_value.rs
Comment (sheepduke) [reply src/evaluation/field_value.rs:147]: The `custom_fields.get()` function returns an `Option` that needs to be unwrapped. The else branch is to handle case where the returned `EvaluationContextFieldValue` is not a `Float` as desired.

File: src/evaluation/context.rs
Comment (sheepduke) [reply src/evaluation/context.rs:39]: The `with_custom_field` takes a `mut Self` and returns a `Self`. It is designed to be chained. But if the client code has already constructed an instance, calling `with_custom_field` on that instance will take the ownership. For example:

```rust
let context = EvaluationContext::default();
context.with_custom_field(...);

// You cannot use context anymore because its ownership has been taken by the `with_xx` function.
// In order to do so, you have to say:
// context = context.with_custom_field(...);
```

I think it is not so convenient considering the custom field might be added lately, so both ways are provided.

You can check [derive_builder's build patterns](https://docs.rs/derive_builder/latest/derive_builder/#builder-patterns) for more details about tradeoffs among different ways.

File: src/api/api.rs
Comment (sheepduke) [reply src/api/api.rs:10]: Ah, you are right. According to the [spec](https://openfeature.dev/specification/sections/flag-evaluation#16-shutdown), we need to define a shutdown function. I need to think of that. I will leave this to the following iterations. :-)

Comment (sheepduke) [reply src/api/api.rs:10]: And you are correct that the singleton will not be destructed, which means that the provider's `drop()` function will never be called...

**Comment by sheepduke** [comment]
> I'm biased b/c I wrote this code, but you might like the pattern we used in the java-sdk where each test is annotated with a specification that it covers. It helps uncover which areas we're not covering yet.
> 
> https://github.com/open-feature/java-sdk/blob/main/src/test/java/dev/openfeature/sdk/FlagEvaluationSpecTest.java#L69

@justinabrahms My apologies. Your request changes have been overwritten by my force-update to resolve the signing issue. :-( 

Could you please kindly re-request it, or let me know what have you changed?

I am rather curious that how did you add the specification to the tests.

File: src/api/api.rs
Comment (sheepduke) [reply src/api/api.rs]: > If we can keep it private so only tests can manually create it that would probably be best. No reason to work outside the API. But if there is a use case please correct me here 🙂

Good idea. Fixed.

**Review by justinabrahms — CHANGES_REQUESTED** [review]
The main thing we need to change is the renovate file being deleted.

As for spec tests, there's a python script in the repo which parses them and validates. Something for later, but you could probably make your life easier by commenting the relevant spec until we get something statically verifiable.

**Comment by sheepduke** [comment]
> The main thing we need to change is the renovate file being deleted.
> 
> As for spec tests, there's a python script in the repo which parses them and validates. Something for later, but you could probably make your life easier by commenting the relevant spec until we get something statically verifiable.

Thanks for sharing the info. I have added this JSON file back and created an issue to track it: 
https://github.com/open-feature/rust-sdk/issues/16

I have re-requested your review. :-)

**Review by justinabrahms — APPROVED** [review]

**Comment by justinabrahms** [comment]
Ahh. Bit of confusion. I'll update the ticket. Renovate keeps our dependencies up to date. The json thing I mentioned was something else.

**Pull request merged**
